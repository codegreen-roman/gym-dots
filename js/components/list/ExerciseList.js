import React from 'react'
import { array, string, func } from 'prop-types'
import glamorous from 'glamorous'

import { ExerciseListRow } from './ExerciseListRow'
import { ExerciseListHeader } from './ExerciseListHeader'

const onOrderChangeClick = () => {} // noop

const _ExerciseList = ({ exercises, title, onOrderChangeClick }) => {

    const Ul = glamorous.ul({
        listStyleType: 'none',
        padding: 0
    })

    return (
        <Ul>
            {exercises.length > 0 ? <ExerciseListHeader title={title} /> : null }
            {exercises.map((exercise, idx) => (
                <ExerciseListRow onRowClick={() => onOrderChangeClick(exercise)} key={idx} exercise={exercise} />
            ))}
        </Ul>
    )
}

_ExerciseList.defaultProps = {
    onOrderChangeClick
}

_ExerciseList.propTypes = {
    exercises: array.isRequired,
    title: string.isRequired,
    onOrderChangeClick: func.isRequired
}

export const ExerciseList = _ExerciseList
