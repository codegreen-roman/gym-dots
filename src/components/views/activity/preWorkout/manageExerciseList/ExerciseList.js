import React from 'react'
import { array, string, func } from 'prop-types'
import {
    exerciseList,
    exerciseListHeader,
    exerciseListHeaderDot,
    exerciseListHeaderText
} from './ExerciseList.glamor'
import { ExerciseListRow } from './ExerciseListRow'
import { compose, not, map, isEmpty } from 'ramda'

export const ExerciseListHeader = ({title}) => {
    return (
        <div data-test='list-title' {...exerciseListHeader}>
            <div {...exerciseListHeaderDot} />
            <div {...exerciseListHeaderText}>{title}</div>
        </div>
    )
}

ExerciseListHeader.propTypes = {
    title: string.isRequired
}

const onOrderChangeClick = () => {} // noop

const _ExerciseList = ({ list, title, onOrderChangeClick }) => {
    const row = (item) => (
        <ExerciseListRow key={item.exerciseKey} {...item} onRowClick={() => onOrderChangeClick(item)} />
    )
    const renderRows = () => map(row, list)
    const notEmpty = compose(not, isEmpty)

    return (notEmpty(list) ?
        (<div>
            <ExerciseListHeader title={title} />
            <ul data-test='list' {...exerciseList}>
                {renderRows()}
            </ul>
        </div>) : null
    )
}

_ExerciseList.defaultProps = {
    onOrderChangeClick
}

_ExerciseList.propTypes = {
    list: array.isRequired,
    title: string.isRequired,
    onOrderChangeClick: func
}

export const ExerciseList = _ExerciseList
