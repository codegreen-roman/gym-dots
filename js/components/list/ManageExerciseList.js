import React, { Component } from 'react'
import { ExerciseList } from './ExerciseList'
import { connect } from 'react-redux'
import { compose } from 'ramda'
import { object, func } from 'prop-types'
import { fetchExercises, exercisesOrderChange } from '../../state/actions'

class _ManageExerciseList extends Component {
    constructor(props) {
        super(props)
    }

    componentWillMount() {
        this.props.fetchExercises()
    }

    render() {
        const { exercises: { upcoming, completed, skipped }, onOrderChange } = this.props
        return (
            <div>
                <ExerciseList exercises={upcoming} title='upcoming' onOrderChangeClick={onOrderChange} />
                <ExerciseList exercises={completed} title='completed' />
                <ExerciseList exercises={skipped} title='skipped' />
            </div>
        )
    }
}

_ManageExerciseList.propTypes = {
    exercises: object.isRequired,
    fetchExercises: func.isRequired,
    onOrderChange: func.isRequired
}

const mapStateToProps = ({ exercises, sessionId, name }) => {
    return {
        exercises,
        sessionId,
        name
    }
}

const mapActionsToProps = dispatch => {
    return {
        fetchExercises: compose(dispatch, fetchExercises),
        onOrderChange: compose(dispatch, exercisesOrderChange)
    }
}

export { _ManageExerciseList }
export const ManageExerciseList = connect(mapStateToProps, mapActionsToProps)(_ManageExerciseList)
