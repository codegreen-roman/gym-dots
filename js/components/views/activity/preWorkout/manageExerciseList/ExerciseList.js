import React from 'react'
import { isEmpty } from 'ramda'
import { array, string, func } from 'prop-types'
import { exerciseList, exerciseListHeader } from './ExerciseList.glamor'
import { ExerciseListRow } from './ExerciseListRow'

const onOrderChangeClick = () => {} // noop

const _ExerciseList = ({ list, title, onOrderChangeClick }) => {
    if (isEmpty(list)) {
        return null
    }
    return (
        <div>
            <span data-test='list-title' {...exerciseListHeader}>
                {title}
            </span>
            <ul data-test='list' {...exerciseList}>
                {list.map((itm, idx) =>
                    <ExerciseListRow
                        key={idx}
                        {...itm}
                        onRowClick={() => onOrderChangeClick(itm)}
                    />
                )}
            </ul>
        </div>
    )
}

_ExerciseList.defaultProps = {
    onOrderChangeClick
}

_ExerciseList.propTypes = {
    list: array.isRequired,
    title: string.isRequired,
    onOrderChangeClick: func
}

export const ExerciseList = _ExerciseList
