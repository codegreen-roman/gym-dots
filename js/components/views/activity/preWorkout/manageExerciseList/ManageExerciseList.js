import React, { Component } from 'react'
import { ExerciseList } from './ExerciseList'
import { connect } from 'react-redux'
import { compose, all, isEmpty, not, isNil } from 'ramda'
import { func, bool, string, array } from 'prop-types'
import { exercisesOrderChange } from '../../../../../state/actions/index'

const notMissing = compose(not, isNil)
const allListsEmpty = all(isEmpty)

class _ManageExerciseList extends Component {

    congratulateUser() {
        const { sessionDone, sessionKey } = this.props

        if (sessionDone) {
            return (<h2>Your session {sessionKey} is COMPLETED!</h2>)
        }

        return null
    }

    render() {
        const { upcoming, completed, skipped, onOrderChange, } = this.props

        return (
            <div>
                {this.congratulateUser()}
                <span>{isEmpty(upcoming) ? 'You got no exercises' : null}</span>
                <ExerciseList key='upcoming' list={upcoming} title='upcoming' onOrderChangeClick={onOrderChange} />
                <ExerciseList key='completed' list={completed} title='completed' />
                <ExerciseList key='skipped' list={skipped} title='skipped' />
            </div>
        )
    }
}

_ManageExerciseList.propTypes = {
    upcoming: array.isRequired,
    completed: array.isRequired,
    skipped: array.isRequired,
    onOrderChange: func.isRequired,
    sessionKey: string,
    sessionDone: bool.isRequired
}

const mapStateToProps = ({ exercises: { upcoming, completed, skipped, sessionKey, name } }) => ({
    upcoming,
    completed,
    skipped,
    name,
    sessionKey,
    sessionDone: allListsEmpty([upcoming, skipped]) && notMissing(sessionKey)
})

const mapActionsToProps = dispatch => {
    return {
        onOrderChange: compose(dispatch, exercisesOrderChange)
    }
}

export { _ManageExerciseList }
export const ManageExerciseList = connect(mapStateToProps, mapActionsToProps)(_ManageExerciseList)
