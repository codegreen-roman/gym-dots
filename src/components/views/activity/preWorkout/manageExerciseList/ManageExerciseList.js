import React, { Component } from 'react'
import { ExerciseList } from './ExerciseList'
import { connect } from 'react-redux'
import { compose, isEmpty } from 'ramda'
import { toWritableResults, allListsEmpty, notMissing } from './ManageExerciseList.helper'
import { func, bool, string, array } from 'prop-types'
import { exercisesOrderChange, saveExercisesResults } from '../../../../../state/actions'

class _ManageExerciseList extends Component {

    componentDidMount() {
        const { sessionDone, sessionKey, completed, saveResults } = this.props
        if (sessionDone) {

            saveResults({
                [sessionKey]: toWritableResults(completed)
            })
        }
    }

    congratulateUser() {
        const { sessionDone, sessionKey } = this.props

        if (sessionDone) {
            return (<h2>Your session {sessionKey} is COMPLETED!</h2>)
        }

        return null
    }

    render() {
        const { upcoming, completed, skipped, onOrderChange, } = this.props

        return (
            <div>
                {this.congratulateUser()}
                <span>{isEmpty(upcoming) ? 'You got no exercises yet' : null}</span>
                <ExerciseList key='upcoming' list={upcoming} title='upcoming' onOrderChangeClick={onOrderChange} />
                <ExerciseList key='completed' list={completed} title='completed' />
                <ExerciseList key='skipped' list={skipped} title='skipped' />
            </div>
        )
    }
}

_ManageExerciseList.propTypes = {
    upcoming: array.isRequired,
    completed: array.isRequired,
    skipped: array.isRequired,
    onOrderChange: func.isRequired,
    saveResults: func.isRequired,
    sessionKey: string,
    sessionDone: bool.isRequired
}

const mapStateToProps = ({ exercises: { upcoming, completed, skipped, sessionKey, name } }) => ({
    upcoming,
    completed,
    skipped,
    name,
    sessionKey,
    sessionDone: allListsEmpty([upcoming, skipped]) && notMissing(sessionKey)
})

const mapActionsToProps = dispatch => {
    return {
        saveResults: compose(dispatch, saveExercisesResults),
        onOrderChange: compose(dispatch, exercisesOrderChange)
    }
}

export { _ManageExerciseList }
export const ManageExerciseList = connect(mapStateToProps, mapActionsToProps)(_ManageExerciseList)
