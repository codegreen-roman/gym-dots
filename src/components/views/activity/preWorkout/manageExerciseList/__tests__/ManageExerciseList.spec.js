import React from 'react'
import { shallow, mount } from 'enzyme'
import { _ManageExerciseList as ManageExerciseList } from '../ManageExerciseList'
import { ExerciseList } from '../ExerciseList'

const props = {
    sessionKey: '',
    name: '',
    sessionDone: false,
    upcoming: [
        {
            exerciseKey: '',
            restTime: 90,
            name: 'Push-ups',
            sets: 5,
            reps: 20,
            weight: 0,
            results: []
        },
        {
            exerciseKey: '',
            restTime: 30,
            name: 'Australian pull-ups',
            sets: 5,
            reps: 12,
            weight: 0,
            results: []
        }
    ],
    skipped: [],
    completed: []
}

const setup = props => {
    const actions = {
        onOrderChange: jest.fn(),
        saveResults: jest.fn(),
    }

    const component = shallow(<ManageExerciseList {...props} {...actions} />)

    return {
        component,
        list: component.find(ExerciseList),
        actions,
        props,
        mounted: mount(<ManageExerciseList {...props} {...actions} />)
    }
}

describe('ManageExerciseList component', () => {

    describe('When one list have exercises others should return null', () => {
        it('should render one list and other are empty', () => {
            const { list } = setup(props)
            expect(list.at(0).props().list.length).toBe(2)
            expect(list.at(1).props().list.length).toBe(0)
            expect(list.at(2).props().list.length).toBe(0)
        })
    })

    describe('No exercises case', () => {
        it('should show \'You got no exercises yet\'', () => {
            const { component } = setup({
                ...props,
                upcoming: []
            })

            expect(component).toMatchSnapshot()
        })
    })

    describe('Session is completed', () => {

        it('should show the completed message', () => {
            const { component } = setup({
                ...props,
                sessionKey: 'thisIsUniqueSessionKey',
                sessionDone: true,
                upcoming: []
            })

            expect(component).toMatchSnapshot()
        })

        it('should call the saveResults method', () => {
            const { actions, mounted } = setup({
                ...props,
                sessionKey: 'thisIsUniqueSessionKey',
                sessionDone: true,
                completed: [
                    {
                        exerciseKey: 'key1',
                        restTime: 90,
                        name: 'Push-ups',
                        sets: 5,
                        reps: 20,
                        weight: 0,
                        allDone: true
                    },
                    {
                        exerciseKey: 'key2',
                        restTime: 30,
                        name: 'Australian pull-ups',
                        sets: 5,
                        reps: 12,
                        weight: 0,
                        allDone: false
                    }
                ]
            })

            mounted.update()
            expect(actions.saveResults).toHaveBeenCalled()
            expect(actions.saveResults).toHaveBeenLastCalledWith(
                {
                    'thisIsUniqueSessionKey': {
                        'key1': true,
                        'key2': false
                    }
                }
            )
        })
    })
})
