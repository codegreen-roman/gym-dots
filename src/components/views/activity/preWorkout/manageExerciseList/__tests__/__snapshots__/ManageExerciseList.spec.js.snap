// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageExerciseList component No exercises case should show 'You got no exercises yet' 1`] = `
<div>
  <span>
    You got no exercises yet
  </span>
  <_ExerciseList
    list={Array []}
    onOrderChangeClick={[Function]}
    title="upcoming"
  />
  <_ExerciseList
    list={Array []}
    onOrderChangeClick={[Function]}
    title="completed"
  />
  <_ExerciseList
    list={Array []}
    onOrderChangeClick={[Function]}
    title="skipped"
  />
</div>
`;

exports[`ManageExerciseList component Session is completed should show the completed message 1`] = `
<div>
  <h2>
    Your session 
    thisIsUniqueSessionKey
     is COMPLETED!
  </h2>
  <span>
    You got no exercises yet
  </span>
  <_ExerciseList
    list={Array []}
    onOrderChangeClick={[Function]}
    title="upcoming"
  />
  <_ExerciseList
    list={Array []}
    onOrderChangeClick={[Function]}
    title="completed"
  />
  <_ExerciseList
    list={Array []}
    onOrderChangeClick={[Function]}
    title="skipped"
  />
</div>
`;
